### PR TITLE
SUP-636 IA-3246 fix bug for dataproc and add test

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/SUP-636

This was an oversight when I was doing the fix last time, somehow I thought dataproc wasn't be affected, which was not true.

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeGceSpec.scala [hunk 1]
```diff
@@ -141,6 +141,7 @@ class RuntimeGceSpec
   private def testStartupScripts(project: GoogleProject, image: Option[ContainerImage] = None): IO[Unit] = {
     dependencies.use { deps =>
       implicit val client = deps.httpClient
+      val checkJupyerSetup = !image.contains(LeonardoConfig.Leonardo.rstudioBioconductorImage)
```
Comment (xuscrbw) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeGceSpec.scala]: missing T in Jupyter

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeGceSpec.scala [hunk 2]
```diff
@@ -215,7 +216,7 @@ class RuntimeGceSpec
 
         // stop/start the runtime
         _ <- IO(stopAndMonitorRuntime(runtime.googleProject, runtime.clusterName))
-        _ <- IO(startAndMonitorRuntime(runtime.googleProject, runtime.clusterName))
+        _ <- IO(startAndMonitorRuntime(runtime.googleProject, runtime.clusterName, checkJupyerSetup))
```
Comment (xuscrbw) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeGceSpec.scala]: missing T in Jupyter

File: http/src/main/resources/init-resources/startup.sh [hunk 3]
```diff
@@ -206,13 +206,23 @@ else
         # what was previously under $HOME will now appear in new $HOME as well
         docker exec $JUPYTER_SERVER_NAME /bin/bash -c "[ -d $JUPYTER_USER_HOME/notebooks ] && [ ! -d $JUPYTER_USER_HOME/notebooks/.jupyter ] && rsync -avr --progress --exclude notebooks . $JUPYTER_USER_HOME/notebooks || true"
 
-        # Make sure when runtimes restarts, they'll get a new version of jupyter docker compose file
-        $GSUTIL_CMD cp gs://${INIT_BUCKET_NAME}/`basename ${JUPYTER_DOCKER_COMPOSE}` $JUPYTER_DOCKER_COMPOSE
+        NEED_MIGRATE=$(docker exec $JUPYTER_SERVER_NAME /bin/bash -c "[ -d $JUPYTER_USER_HOME/notebooks ] && echo 'true' || echo 'false'")
```
Comment (calbach) [inline http/src/main/resources/init-resources/startup.sh:207]: Maybe add a  comment - why does the presence of this directory indicate that it needs migration?

File: http/src/main/resources/init-resources/startup.sh [hunk 4]
```diff
@@ -206,13 +206,23 @@ else
         # what was previously under $HOME will now appear in new $HOME as well
         docker exec $JUPYTER_SERVER_NAME /bin/bash -c "[ -d $JUPYTER_USER_HOME/notebooks ] && [ ! -d $JUPYTER_USER_HOME/notebooks/.jupyter ] && rsync -avr --progress --exclude notebooks . $JUPYTER_USER_HOME/notebooks || true"
```
Comment (calbach) [inline http/src/main/resources/init-resources/startup.sh]: Should this line now be moved into the NEEDS_MIGRATE block with a simplified predicate?

**Review by calbach — COMMENTED** [review]
I only reviewed the bash changes

File: http/src/main/resources/init-resources/startup.sh (hunk 3 above)
Comment (Qi77Qi) [reply http/src/main/resources/init-resources/startup.sh:207]: think [this](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/init-resources/startup.sh#L128) is good enough?

Comment (calbach) [reply http/src/main/resources/init-resources/startup.sh:207]: Oops, yep -missed that was above

**Comment by calbach** [comment]
Bash looks good.. I'll let @xuscrbw do the approvals since I didn't review any of the scala changes

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala [hunk 5]
```diff
@@ -409,7 +409,7 @@ object Boot extends IOApp {
         F.delay(CaffeineCache[F, UserEmailAndProject, Option[String]](underlyingPetTokenCache))
       )(s => F.delay(s.close))
 
-      samDao = HttpSamDAO[F](Retry(retryPolicy)(httpClient), httpSamDaoConfig, petTokenCache)
+      samDao = HttpSamDAO[F](httpClientWithRetryAndLogging, httpSamDaoConfig, petTokenCache)
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala:412]: added logging becuz I'm seeing lots of `"Max wait queue for limit of 256 for https://sam.dsde-prod.broadinstitute.org:443 reached, not scheduling."` in prod

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala [hunk 6]
```diff
@@ -223,15 +223,6 @@ class ProxyRoutes(proxyService: ProxyService, corsSupport: CorsSupport, refererC
               .fromFuture(IO(request.entity.discardBytes().future))
         }
       resp <- ctx.span.fold(apiCall)(span => spanResource[IO](span, "proxyApp").use(_ => apiCall))
-
-      _ <- if (resp.status.isSuccess()) {
-        metrics.incrementCounter("proxyRequest",
-                                 tags = Map("result" -> "success", "action" -> "appRequest", "tool" -> s"${appName}"))
-      } else {
-        metrics.incrementCounter("proxyRequest",
-                                 tags = Map("result" -> "failure", "action" -> "appRequest", "tool" -> s"${appName}"))
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala:232]: this `appName` is wrong...metrics are not aggregated properly this way. Hence did some refactoring to fix this

**Review by xuscrbw — APPROVED** [review]
lgtm

File: http/src/main/resources/reference.conf [hunk 7]
```diff
@@ -573,14 +573,14 @@ clusterFiles {
 # so we need to ensure we don't use stale cache entries.
 runtimeDnsCache {
   cacheExpiryTime = 5 seconds
-  cacheMaxSize = 1000
+  cacheMaxSize = 100
```
Comment (Qi77Qi) [inline http/src/main/resources/reference.conf:576]: Looked at [prod metrics](https://grafana.dsp-devops.broadinstitute.org/d/470LS8A7k/leonardo?orgId=1) hence adjusting size a bit

**Pull request merged**